### PR TITLE
feat(admin): exam mode configuration (#180)

### LINKED ISSUE #260: Exam mode: enforce full-submission guard in batch_submit_quiz
## Context
CodeRabbit flagged on PR #258 that `batch_submit_quiz` accepts partial submissions (1/N answers) and permanently completes the session. This is **intentional for study mode** (since migration 017) but **must be blocked for exam mode**.

## Current behavior (correct for study mode)
- Students can skip questions and submit only what they answered
- Score = `correct / answered` (not `correct / total`)
- Session completes with `answered_count < total_questions`

## Required for exam mode (#180)
When exam mode is implemented, add a mode-aware guard:
```sql
IF v_mode = 'exam' AND jsonb_array_length(p_answers) <> v_total THEN
  RAISE EXCEPTION 'exam mode requires all questions answered';
END IF;
```

This requires:
1. A `mode` column on `quiz_sessions` (or in `config` JSONB)
2. The guard in `batch_submit_quiz` before the answer loop
3. Red-team spec to verify the guard

## Priority
Deferred until Exam Mode (#180) is implemented.

### PR DESCRIPTION
**Pull request by okpilot**
## Summary
- Admin UI at `/app/admin/exam-config` to configure practice exam parameters per EASA subject
- New `exam_configs` + `exam_config_distributions` tables with admin-only RLS
- `start_exam_session` RPC: random question selection per topic distribution
- `batch_submit_quiz` exam guard (#260): blocks partial submission in exam mode, computes pass/fail
- Server-side time limit enforcement (30s grace period)
- Duplicate active exam session prevention (EXISTS guard + partial unique index)
- Zod schemas with distribution sum validation + 26 tests
- Next.js 16.2.2 → 16.2.3 (GHSA-q4gf-8mx6-v5v3)

## Related
- Closes #260 (exam full-submission guard)
- Part of #180 (exam mode — admin config is PR1, student UX is PR2 #514)
- Follow-up: #515 (partial unique index for soft-delete re-creation)

## Test plan
- [x] Visit `/app/admin/exam-config` as admin — see all 9+ subjects listed
- [x] Click a subject → configure total questions, time limit, pass mark
- [x] Set question distribution per topic — verify sum validation works
- [x] Save config → verify it persists on page reload
- [x] Toggle enable/disable — verify validation prevents enabling without complete distribution
- [x] Run `pnpm test` — 2676 tests passing (26 new Zod schema tests)
- [x] Verify type-check passes (`pnpm check-types`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Admin "Exam Config" dashboard to define per-subject exam settings (total questions, time limit, pass mark), topic/subtopic distributions, and enable/disable per subject; RPC to atomically save configs.

* **UI**
  * New navigation item and icon, loading skeletons, subject cards, modal config editor, distribution editor, and numeric input control.

* **Behavior & Server**
  * Distribution-sum validation, single active mock exam per student, server-enforced time limits with auto-submit, guarded session start, and pass/fail persisted.

* **Database & Security**
  * New exam config/distribution tables, quiz_sessions exam columns, soft-delete semantics, RLS rules, partial-unique index, and guarded RPCs.

* **Tests & Docs**
  * New spec/requirements/tasks docs, migrations, types/schemas, and extensive tests for UI, actions, validation, and RPCs.

### CODE REVIEW

File: apps/web/app/app/admin/exam-config/_components/exam-config-page-shell.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      </div>
+
+      {editingSubject && (
+        <ConfigFormDialog
+          subject={editingSubject}
+          open={!!editingSubject}
+          onOpenChange={(open) => {
+            if (!open) setEditingSubject(null)
+          }}
+        />
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/exam-config-page-shell.tsx:33]: Fixed in 23edbce. Changed to `open={true}` (always true inside mount guard) and added `key={editingSubject.id}` to force clean remount when switching subjects.

File: apps/web/app/app/admin/exam-config/_components/num-field.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  return (
+    <div>
+      <label className={labelCls}>
+        {label}
+        <input
+          type="number"
+          min={min}
+          max={max}
+          value={value}
+          onChange={(e) => onChange(Number(e.target.value))}
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/num-field.tsx]: Fixed in 23edbce. Replaced `Number(e.target.value)` with `e.target.valueAsNumber` + `Number.isNaN()` guard + `Math.max(min, Math.min(max, v))` clamp.

File: apps/web/app/app/admin/exam-config/_components/config-form-dialog.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            onClick={handleSubmit}
+            disabled={isPending || !isValid}
+            className="rounded bg-primary px-4 py-2 text-sm font-medium text-primary-foreground hover:bg-primary/90 disabled:opacity-50"
+          >
+            {isPending ? 'Saving...' : 'Save Config'}
+          </button>
+        </div>
+      </div>
+    </div>
+  )
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/config-form-dialog.tsx:142]: Fixed in 23edbce. Replaced hand-rolled `<div>` overlay with shadcn `Dialog`, `DialogContent`, `DialogHeader`, `DialogFooter`. Added `Switch` for the enabled toggle and `Button` for actions. Proper focus trap, ESC handler, and ARIA attributes now come from shadcn.

File: apps/web/app/app/admin/exam-config/actions/toggle-exam-config.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    if (!config) {
+      return { success: false, error: 'Configure exam parameters before enabling' }
+    }
+
+    const { data: distributions } = await supabase
+      .from('exam_config_distributions')
+      .select('question_count')
+      .eq('exam_config_id', config.id)
+
+    const totalDistributed = (distributions ?? []).reduce((sum, d) => sum + d.question_count, 0)
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/actions/toggle-exam-config.ts]: Fixed in 23edbce. Both Supabase reads now destructure `error` with distinct log messages (`configErr`, `distErr`) and return generic error strings to the client.

File: apps/web/app/app/admin/exam-config/actions/upsert-exam-config.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      question_count: d.questionCount,
+    }))
+
+    const { error: insertError } = await supabase.from('exam_config_distributions').insert(rows)
+
+    if (insertError) {
+      console.error('[upsertExamConfig] Insert distributions error:', insertError.message)
+      return { success: false, error: 'Failed to save question distribution' }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/actions/upsert-exam-config.ts]: Fixed in edee5be. The entire multi-table mutation (config upsert + distribution replace) is now wrapped in a single `upsert_exam_config` SECURITY DEFINER RPC — one Postgres transaction, fully atomic. The delete+reinsert of distributions is intentional for this ephemeral table (same precedent as `quiz_drafts`, documented in migration 006 and database.md).

File: docs/database.md [hunk 6]
```diff
@@ -243,12 +243,47 @@ CREATE TABLE quiz_sessions (
...
+CREATE TABLE exam_config_distributions (
+  id                UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+  exam_config_id    UUID NOT NULL REFERENCES exam_configs(id) ON DELETE CASCADE,
+  topic_id          UUID NOT NULL REFERENCES easa_topics(id),
+  subtopic_id       UUID REFERENCES easa_subtopics(id) NULL,
+  question_count    INT NOT NULL CHECK (question_count > 0),
+  UNIQUE (exam_config_id, topic_id, subtopic_id)
+);
+-- RLS: admin-only (students cannot read distributions)
+```
```
[the remark this answers is not in the document]
Comment (okpilot) [reply docs/database.md:289]: Skip — false positive. The student SELECT policy `student_select_exam_configs` exists at migration 001 lines 99-105. The RLS note in database.md is correct.

File: docs/database.md [hunk 7]
```diff
@@ -243,12 +243,47 @@ CREATE TABLE quiz_sessions (
...
+  updated_at        TIMESTAMPTZ NOT NULL DEFAULT now(),
+  deleted_at        TIMESTAMPTZ NULL,
+  UNIQUE (organization_id, subject_id)
+);
+-- RLS: admin full CRUD, students read-only (enabled + non-deleted only)
+```
+
+### exam_config_distributions
+```sql
+-- Question count per topic/subtopic for exam random selection.
```
[the remark this answers is not in the document]
Comment (okpilot) [reply docs/database.md:277]: Fixed in 23edbce. Added proper spacing and formatting around `### exam_configs` and `### exam_config_distributions` headings and fenced SQL blocks.

File: packages/db/migrations/040_start_exam_session_rpc.sql [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+      WHERE q.subject_id = p_subject_id
+        AND q.topic_id = v_dist.topic_id
+        AND (v_dist.subtopic_id IS NULL OR q.subtopic_id = v_dist.subtopic_id)
+        AND q.status = 'active'
+        AND q.deleted_at IS NULL
+        AND q.organization_id = v_org_id
+        AND q.id != ALL(v_selected_ids)  -- prevent duplicates across distributions
+      ORDER BY random()
+      LIMIT v_dist.question_count
+    );
```
[the remark this answers is not in the document]
Comment (okpilot) [reply packages/db/migrations/040_start_exam_session_rpc.sql:76]: Deferred to PR2 (#514). This is an edge case in pool sizing when broad topic rows consume questions that later subtopic rows need. Added to #514 body with detailed description and proposed fix (`NULLS LAST` or pre-allocate subtopic-specific questions first).

File: apps/web/app/app/admin/exam-config/actions/upsert-exam-config.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    parsed.data
+
+  // Check if config exists
+  const { data: existing } = await supabase
+    .from('exam_configs')
+    .select('id')
+    .eq('organization_id', organizationId)
+    .eq('subject_id', subjectId)
+    .is('deleted_at', null)
+    .maybeSingle()
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/actions/upsert-exam-config.ts]: Fixed in 23edbce (partial unique index migration `20260411000007`) + edee5be (RPC extraction). The full UNIQUE constraint was replaced with a partial unique index `WHERE deleted_at IS NULL`, so soft-deleted configs no longer block re-creation. The upsert logic was then extracted into a SECURITY DEFINER RPC for ACID compliance.

File: packages/db/migrations/042_exam_config_security_fixes.sql [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+
+  RETURN jsonb_build_object(
+    'session_id', v_session_id,
+    'question_ids', to_jsonb(v_selected_ids),
+    'time_limit_seconds', v_time_limit,
+    'total_questions', v_total_questions,
+    'pass_mark', v_pass_mark
+  );
+END;
+$$;
```
[the remark this answers is not in the document]
Comment (okpilot) [reply packages/db/migrations/042_exam_config_security_fixes.sql:433]: Skip — immutable migration. This migration is already committed and applied to the database. Per project rules, existing migrations cannot be modified. Future migrations will stay under the 300-line limit.

File: supabase/migrations/20260411000005_exam_config_security_fixes.sql [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+    );
+  END IF;
+
+  -- *** SERVER-SIDE TIME LIMIT ENFORCEMENT ***
+  -- Prevents bypassing client-side countdown by delaying submission
+  -- Grace period of 30 seconds accounts for network latency
+  IF v_time_limit IS NOT NULL AND v_started_at IS NOT NULL THEN
+    IF now() > v_started_at + (v_time_limit + 30) * interval '1 second' THEN
+      RAISE EXCEPTION 'exam time limit exceeded';
+    END IF;
```
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000005_exam_config_security_fixes.sql:111]: Duplicate of finding on packages/db copy. Deferred to PR2 (#514) — see reply above.

File: packages/db/migrations/042_exam_config_security_fixes.sql [hunk 12]
```diff
@@ -0,0 +1,433 @@
...
+    );
+  END IF;
+
+  -- *** SERVER-SIDE TIME LIMIT ENFORCEMENT ***
+  -- Prevents bypassing client-side countdown by delaying submission
+  -- Grace period of 30 seconds accounts for network latency
+  IF v_time_limit IS NOT NULL AND v_started_at IS NOT NULL THEN
+    IF now() > v_started_at + (v_time_limit + 30) * interval '1 second' THEN
+      RAISE EXCEPTION 'exam time limit exceeded';
+    END IF;
```
[the remark this answers is not in the document]
Comment (okpilot) [reply packages/db/migrations/042_exam_config_security_fixes.sql:111]: Deferred to PR2 (#514). The RAISE EXCEPTION rolls back the transaction, leaving `ended_at = NULL` which blocks future sessions via the partial unique index. Added to #514 body — fix should mark session as ended/failed instead of raising.

File: supabase/migrations/20260411000003_start_exam_session_rpc.sql [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+      WHERE q.subject_id = p_subject_id
+        AND q.topic_id = v_dist.topic_id
+        AND (v_dist.subtopic_id IS NULL OR q.subtopic_id = v_dist.subtopic_id)
+        AND q.status = 'active'
+        AND q.deleted_at IS NULL
+        AND q.organization_id = v_org_id
+        AND q.id != ALL(v_selected_ids)  -- prevent duplicates across distributions
+      ORDER BY random()
+      LIMIT v_dist.question_count
+    );
```
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000003_start_exam_session_rpc.sql:76]: Duplicate of finding on packages/db copy. Deferred to PR2 (#514) — see reply above.

File: apps/web/app/app/admin/exam-config/_components/num-field.tsx [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        <input
+          type="number"
+          min={min}
+          max={max}
+          value={value}
+          onChange={(e) => {
+            const v = e.target.valueAsNumber
+            if (Number.isNaN(v)) return
+            onChange(Math.max(min, Math.min(max, v)))
+          }}
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/num-field.tsx:32]: Skipping — all 3 usages of NumField in `config-form-dialog.tsx` use `min=1` (lines 89, 96, 103). With min=1, no single digit gets clamped upward so the per-keystroke clamp works correctly. Moving to onBlur would introduce temporary invalid state in the parent. The bug only manifests with `min >= 10`, which we don't have.

File: apps/web/app/app/admin/exam-config/_components/num-field.test.tsx [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+      const onChange = vi.fn()
+      render(<NumField label="Total" value={10} min={1} max={100} onChange={onChange} />)
+      const input = screen.getByRole('spinbutton')
+
+      await userEvent.clear(input)
+      await userEvent.type(input, '150')
+
+      const lastCall = onChange.mock.calls.at(-1)?.[0] as number
+      expect(lastCall).toBeLessThanOrEqual(100)
+    })
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/num-field.test.tsx]: Fixed in 8556c57. Changed to `toBe(100)` and `toBe(5)` using `fireEvent.change` to bypass the controlled-component re-render issue with `userEvent.clear + type`.

File: apps/web/app/app/admin/exam-config/actions/toggle-exam-config.ts [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  const { data, error } = await supabase
+    .from('exam_configs')
+    .update({
+      enabled,
+      updated_at: new Date().toISOString(),
+    })
+    .eq('organization_id', organizationId)
+    .eq('subject_id', subjectId)
+    .is('deleted_at', null)
+    .select('id')
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/actions/toggle-exam-config.ts:63]: Deferred to PR2 — tracked in #522. Valid TOCTOU concern but single-tenant admin page makes it extremely low probability. The fix requires a new SECURITY DEFINER RPC with `SELECT FOR UPDATE`, which is in scope for #514.

File: supabase/migrations/20260411000008_upsert_exam_config_rpc.sql [hunk 17]
```diff
@@ -0,0 +1,97 @@
+-- upsert_exam_config: atomically upserts an exam configuration and replaces
+-- its topic/subtopic question distribution in a single transaction.
+--
+-- Parameters:
+--   p_subject_id:         EASA subject UUID
+--   p_enabled:            whether exam mode is active for this subject
+--   p_total_questions:    number of questions in the exam
+--   p_time_limit_seconds: exam duration in seconds
+--   p_pass_mark:          pass percentage (1-100)
+--   p_distributions:      JSONB array of {topicId, subtopicId, questionCount}
```
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000008_upsert_exam_config_rpc.sql]: Fixed in bd0d55a. Changed comment from `{topicId, subtopicId, questionCount}` to `{topic_id, subtopic_id, question_count}` in both migration files.

File: docs/database.md [hunk 18]
```diff
@@ -243,12 +243,51 @@ CREATE TABLE quiz_sessions (
...
+CREATE TABLE exam_config_distributions (
+  id                UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+  exam_config_id    UUID NOT NULL REFERENCES exam_configs(id) ON DELETE CASCADE,
+  topic_id          UUID NOT NULL REFERENCES easa_topics(id),
+  subtopic_id       UUID REFERENCES easa_subtopics(id) NULL,
+  question_count    INT NOT NULL CHECK (question_count > 0),
+  UNIQUE (exam_config_id, topic_id, subtopic_id)
+);
+-- RLS: admin-only (students cannot read distributions)
+```
```
[the remark this answers is not in the document]
Comment (okpilot) [reply docs/database.md:289]: Fixed in edbfe3d. Added both tables to §3 soft-delete matrix: `exam_configs` (soft-delete) and `exam_config_distributions` (hard DELETE approved exception — two paths: CASCADE from parent + direct DELETE in `upsert_exam_config` RPC).

File: supabase/migrations/20260411000008_upsert_exam_config_rpc.sql [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+-- Security: SECURITY DEFINER with admin role check + org scoping
+
+CREATE OR REPLACE FUNCTION upsert_exam_config(
+  p_subject_id         uuid,
+  p_enabled            boolean,
+  p_total_questions    int,
+  p_time_limit_seconds int,
+  p_pass_mark          int,
+  p_distributions      jsonb
+) RETURNS uuid
```
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000008_upsert_exam_config_rpc.sql:23]: Deferred to PR2 — tracked in #521. Valid defense-in-depth gap. The Zod schema enforces `distributions.min(1)` and sum equality at the Server Action boundary, but the RPC itself should also validate. Fix requires a new migration with the validation logic added to `upsert_exam_config`.

File: apps/web/app/app/admin/exam-config/_components/exam-config-page-shell.test.tsx [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+    it('opens the dialog for the correct subject when a card edit is triggered', async () => {
+      render(<ExamConfigPageShell subjects={[SUBJECT_1, SUBJECT_2]} />)
+
+      await userEvent.click(screen.getByTestId('card-subj-2'))
+
+      const dialog = screen.getByTestId('dialog')
+      expect(dialog.getAttribute('data-subject-id')).toBe('subj-2')
+    })
+
+    it('closes the dialog when onOpenChange(false) is called', async () => {
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/exam-config-page-shell.test.tsx]: Fixed in 35a9c73. Renamed to "closes the dialog when the close button is clicked" — describes user behavior, not callback internals.

File: apps/web/app/app/admin/exam-config/_components/subject-config-card.test.tsx [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+      const onEdit = vi.fn()
+      render(<SubjectConfigCard subject={baseSubject} onEdit={onEdit} />)
+      await userEvent.click(screen.getByText('Not configured'))
+      expect(onEdit).toHaveBeenCalledTimes(1)
+    })
+  })
+
+  describe('toggle button — no config path', () => {
+    it('calls onEdit instead of toggleExamConfig when there is no config', async () => {
+      const onEdit = vi.fn()
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/subject-config-card.test.tsx]: Fixed in 35a9c73. Renamed to "opens edit flow when clicking an unconfigured subject card" and "toggles exam mode when clicking the status button" — behavior-first, no internal names.

File: apps/web/app/app/admin/exam-config/queries.test.ts [hunk 22]
```diff
@@ -0,0 +1,408 @@
...
+
+      await expect(getExamConfigData()).rejects.toThrow('distributions DB error')
+    })
+
+    it('throws when the questions query fails', async () => {
+      mockAdmin()
+      buildTableMocks({ questionsError: { message: 'questions DB error' } })
+
+      await expect(getExamConfigData()).rejects.toThrow('questions DB error')
+    })
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/queries.test.ts:398]: Skipping — cosmetic nitpick. The 3-line arrange pattern is clear and readable as-is. Test files have relaxed line limits per code-style.md, and extracting a helper here doesn't improve clarity.

File: apps/web/app/app/admin/exam-config/queries.test.ts [hunk 23]
```diff
@@ -0,0 +1,408 @@
...
+
+      const result = await getExamConfigData()
+
+      expect(result[0]!.config?.distributions).toEqual([])
+    })
+
+    it('returns subjects with config null when configs data is null', async () => {
+      mockAdmin()
+      buildTableMocks({ configsData: null as unknown as unknown[] })
+
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/queries.test.ts:337]: Skipping — intentional test-only cast. `null as unknown as unknown[]` is used to simulate the edge case where Supabase returns null instead of an array. This is a standard pattern for testing error/edge paths, not production code.

File: supabase/migrations/20260411000008_upsert_exam_config_rpc.sql [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+  ELSE
+    -- Insert new
+    INSERT INTO exam_configs (
+      organization_id, subject_id, enabled,
+      total_questions, time_limit_seconds, pass_mark
+    ) VALUES (
+      v_org_id, p_subject_id, p_enabled,
+      p_total_questions, p_time_limit_seconds, p_pass_mark
+    ) RETURNING id INTO v_config_id;
+  END IF;
```
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000008_upsert_exam_config_rpc.sql:77]: Skipping for this PR — already tracked in #521 for PR2. The partial unique index `uq_exam_configs_org_subject_active` catches the race with a constraint violation error (not silent corruption). Admin-only RPC makes concurrent saves near-impossible in practice.

File: apps/web/app/app/admin/exam-config/_components/config-form-dialog.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/config-form-dialog.tsx:142]: Fixed in 23edbce. Replaced hand-rolled overlay with shadcn `Dialog`/`DialogContent`/`DialogHeader`/`DialogFooter` — gives us focus trapping, ESC-to-close, and proper `role="dialog"` semantics for free.

File: apps/web/app/app/admin/exam-config/actions/upsert-exam-config.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/actions/upsert-exam-config.ts]: Fixed in 4a72ca2 + edee5be. Added partial unique index `uq_exam_configs_org_subject_active` (`WHERE deleted_at IS NULL`) so soft-deleted rows are invisible to the constraint. Extracted into `upsert_exam_config()` SECURITY DEFINER RPC for ACID compliance.

File: apps/web/app/app/admin/exam-config/actions/toggle-exam-config.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/actions/toggle-exam-config.ts]: Fixed in 23edbce. Destructured `configErr`/`distErr` with early returns on truthy errors — both log server-side and return generic error messages.

File: apps/web/app/app/admin/exam-config/_components/num-field.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/num-field.tsx]: Fixed in 23edbce. Switched to `valueAsNumber` with a NaN guard (`if (Number.isNaN(v)) return`) and `Math.max(min, Math.min(max, v))` clamping.

File: apps/web/app/app/admin/exam-config/_components/exam-config-page-shell.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/exam-config-page-shell.tsx:33]: Fixed in 23edbce. Changed to `open={true}` inside the mount guard, plus added `key={editingSubject.id}` to reset form state on subject switch.

File: apps/web/app/app/admin/exam-config/actions/upsert-exam-config.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/actions/upsert-exam-config.ts]: Fixed in edee5be. The delete+reinsert now runs inside the `upsert_exam_config()` SECURITY DEFINER RPC, giving us a single-transaction ACID guarantee. `exam_config_distributions` is intentionally ephemeral — no `deleted_at`, audit trail via parent `exam_configs.updated_at`.

File: docs/database.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply docs/database.md:289]: The `student_select_exam_configs` policy is real — verified in migration 042. The docs note is accurate as written.

File: packages/db/migrations/042_exam_config_security_fixes.sql (hunk 12 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply packages/db/migrations/042_exam_config_security_fixes.sql:111]: Deferred to PR2 (#514). The fix (mark session as `ended_at = now()` / failed instead of raising) will be implemented with the student countdown timer.

File: docs/database.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply docs/database.md:277]: Fixed in edbfe3d. Added blank lines around headings and fenced SQL blocks.

File: packages/db/migrations/040_start_exam_session_rpc.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply packages/db/migrations/040_start_exam_session_rpc.sql:76]: Deferred to PR2 (#514). The proposed fix (`NULLS LAST` ordering or pre-allocating subtopic-specific rows first) will be implemented alongside the student exam flow.

File: supabase/migrations/20260411000003_start_exam_session_rpc.sql (hunk 13 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000003_start_exam_session_rpc.sql:76]: Deferred to PR2 (#514) — duplicate of the packages/db migration comment. Same fix planned.

File: packages/db/migrations/042_exam_config_security_fixes.sql (hunk 10 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply packages/db/migrations/042_exam_config_security_fixes.sql:433]: Migration 042 is immutable — the 300-line cap applies to future migrations only. Cannot split a committed migration.

File: supabase/migrations/20260411000005_exam_config_security_fixes.sql (hunk 11 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000005_exam_config_security_fixes.sql:111]: Deferred to PR2 (#514) — duplicate of the packages/db migration comment. Same fix planned.

File: apps/web/app/app/admin/exam-config/_components/num-field.test.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/num-field.test.tsx]: Fixed in 8556c57. Switched to exact boundary assertions (`toBe(100)` / `toBe(5)`) and `fireEvent.change` for controlled-component compatibility.

File: docs/database.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply docs/database.md:289]: Fixed in edbfe3d. Added both tables to the soft-delete matrix with the two deletion paths (CASCADE from parent + direct DELETE in the RPC) documented.

File: apps/web/app/app/admin/exam-config/_components/num-field.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/num-field.tsx:32]: Acknowledged — with all callers using `min=1`, no in-progress digit gets clamped. Current behavior is correct for our use cases.

File: apps/web/app/app/admin/exam-config/actions/toggle-exam-config.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/actions/toggle-exam-config.ts:63]: Deferred to #522. Will be addressed in PR2.

File: supabase/migrations/20260411000008_upsert_exam_config_rpc.sql (hunk 19 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000008_upsert_exam_config_rpc.sql:23]: Deferred to #521. The Zod boundary covers the happy path; the RPC-level `p_distributions` array/sum check is tracked as defense-in-depth hardening for PR2.

File: supabase/migrations/20260411000008_upsert_exam_config_rpc.sql (hunk 17 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000008_upsert_exam_config_rpc.sql]: Fixed in bd0d55a + aac3e61. Both supabase and packages/db migrations now document the correct snake_case keys `{topic_id, subtopic_id, question_count}`.

File: apps/web/app/app/admin/exam-config/_components/exam-config-page-shell.test.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/exam-config-page-shell.test.tsx]: Fixed in 35a9c73. Renamed to behavior-first wording: "closes the dialog when the close button is clicked".

File: apps/web/app/app/admin/exam-config/_components/subject-config-card.test.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/subject-config-card.test.tsx]: Fixed in 35a9c73. Both titles now describe observable behavior rather than internal function names.

File: apps/web/app/app/admin/exam-config/_components/subject-config-card.test.tsx [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+
+      render(<SubjectConfigCard subject={subjectWithConfig} onEdit={vi.fn()} />)
+      await userEvent.click(screen.getByRole('button', { name: 'Enabled' }))
+
+      await waitFor(() => {
+        expect(toast.error).toHaveBeenCalled()
+      })
+      expect(toast.success).not.toHaveBeenCalled()
+    })
+  })
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/subject-config-card.test.tsx:214]: Fixed in ff3160f. Added `shows toast.error with fallback message when toggleExamConfig throws` test — mocks the action to reject and asserts the fallback toast message.

File: apps/web/app/app/admin/exam-config/queries.test.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/queries.test.ts:398]: Acknowledged — the repetition is intentional for test readability. Each test is self-contained. Won't extract.

File: supabase/migrations/20260411000008_upsert_exam_config_rpc.sql (hunk 24 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply supabase/migrations/20260411000008_upsert_exam_config_rpc.sql:77]: Deferred to #521. The partial unique index `uq_exam_configs_org_subject_active` ensures the race results in a constraint violation rather than silent data corruption. The admin-only single-tenant context makes the window negligible.

File: apps/web/app/app/admin/exam-config/queries.test.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/queries.test.ts:337]: Acknowledged — noted.

File: apps/web/app/app/admin/exam-config/_components/subject-config-card.tsx [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+      const result = await toggleExamConfig({
+        subjectId: subject.id,
+        enabled: !config.enabled,
+      })
+      if (!result.success) {
+        toast.error(result.error)
+      } else {
+        toast.success(config.enabled ? 'Exam mode disabled' : 'Exam mode enabled')
+      }
+    })
```
[the remark this answers is not in the document]
Comment (okpilot) [reply apps/web/app/app/admin/exam-config/_components/subject-config-card.tsx:37]: Fixed in 1395d92. Added try/catch around `toggleExamConfig` call inside `startTransition` with `toast.error('Failed to update exam status')` fallback. Also fixed the same pattern in `config-form-dialog.tsx`. Test coverage added in ff3160f.
